src/Language/Haskell/TH/Lift/Generics/Internal/OuterTypeRep.hs

```diff
@@ -0,0 +1,72 @@
+{-# language CPP #-}
+{-# language FlexibleInstances #-}
+{-# language UndecidableInstances #-}
+{-# language ScopedTypeVariables #-}
+#if __GLASGOW_HASKELL__ >= 708
+{-# language PolyKinds #-}
+#endif
+
+#if __GLASGOW_HASKELL__ < 710
+{-# language OverlappingInstances #-}
+#endif
+
+module Language.Haskell.TH.Lift.Generics.Internal.OuterTypeRep
+  ( OuterTypeable (..)
+  ) where
+import Data.Typeable
+
+-- | A type is @OuterTypable@ if its /type constructor/ is
+-- 'Typeable'. For example, @'Maybe' a@ is only 'Typeable' if
+-- @a@ is 'Typeable', but it is always @OuterTypeable@.
+-- Users should not have to use this class directly; it is
+-- exported for documentation purposes.
+--
+-- === Caution
+--
+-- Before GHC 7.8, only the last seven arguments are properly
+-- supported. So given a type
+--
+-- @
+-- data Foo a b c d e f g h i = ...
+--   deriving (Typeable)
+-- @
+--
+-- @OuterTypeable (Foo a b c d e f g h i)@ requires that @a@
+-- and @b@ are 'Typeable', but the rest of the arguments need
+-- not be.
+class OuterTypeable a where
+  -- | Get the 'TypeRep' corresponding to the outermost constructor
+  -- of a type.
+  getConTR :: proxy a -> TypeRep
+
+#if __GLASGOW_HASKELL__ >= 708
+instance
+# if __GLASGOW_HASKELL__ >= 710
+  {-# OVERLAPPING #-}
+# endif
+  OuterTypeable f => OuterTypeable (f a) where
+  getConTR _ = getConTR (Proxy :: Proxy f)
+instance Typeable a => OuterTypeable a where
+  getConTR = typeRep
+#else
```

Comment:
Yes, we need to use overlapping instances even for GHC >= 7.8. I wanted to use an auxiliary class with a closed type family:

```haskell
type family IsApplication (a :: k) :: Bool where
  IsApplication (f a) = 'True  -- Or IsApplication ((f :: j -> k) a), or whatever.
  IsApplication a = 'False
```

Unfortunately, that sort of thing doesn't seem to be supported until `TypeInType` showed up in GHC 8.0, which is exactly when we stop having any use for it here. I get a mysterious error that looks like

```
    Family instance purports to bind type variable ‘k1’
      but the real LHS (expanding synonyms) is: IsApplication (f a) = ...
    In the equations for closed type family ‘IsApplication’
    In the type family declaration for ‘IsApplication’
```